providers/implementations/keymgmt/template_kmgmt.c

```diff
@@ -0,0 +1,437 @@
+                               const OSSL_PARAM params[])
+{
+    struct template_gen_ctx *gctx = NULL;
+
+    debug_print("gen init called for %p\n", provctx);
+
+    /* perform algorithm type specific sanity checks */
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    if ((gctx = OPENSSL_zalloc(sizeof(*gctx))) != NULL) {
+        gctx->provctx = provctx;
+        gctx->selection = selection;
+    }
+    if (!template_gen_set_params(gctx, params)) {
+        OPENSSL_free(gctx);
+        gctx = NULL;
+    }
+    debug_print("gen init returns %p\n", gctx);
+    return gctx;
+}
+
+static const OSSL_PARAM *template_gen_settable_params(ossl_unused void *genctx,
+                                                      ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_END
+    };
+    return settable;
+}
+
+static void *template_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    struct template_gen_ctx *gctx = (struct template_gen_ctx *)vctx;
+    void *key = NULL;
+
+    debug_print("gen called for %p\n", gctx);
+
+    if (gctx == NULL)
+        goto err;
+
+    /* generate and return new key */
+
+    debug_print("gen returns set %p\n", key);
+    return key;
+err:
+    template_free(key);
+    debug_print("gen returns NULL\n");
+    return NULL;
+}
+
+static void template_gen_cleanup(void *genctx)
+{
+    struct template_gen_ctx *gctx = genctx;
+
+    debug_print("gen cleanup for %p\n", gctx);
+    OPENSSL_free(gctx);
+}
+
+static void *template_dup(const void *vsrckey, int selection)
+{
+    void *dstkey = NULL;
+
+    debug_print("dup called for %p\n", vsrckey);
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    dstkey = template_new(NULL);
+    if (dstkey == NULL)
+        goto err;
+
+    /* populate dstkey from vsrckey material */
+
+    debug_print("dup returns %p\n", dstkey);
+    return dstkey;
+ err:
+    template_free(dstkey);
+    debug_print("dup returns NULL\n");
+    return NULL;
+}
+
+const OSSL_DISPATCH ossl_template_keymgmt_functions[] = {
+    { OSSL_FUNC_KEYMGMT_NEW, (void (*)(void))template_new },
+    { OSSL_FUNC_KEYMGMT_FREE, (void (*)(void))template_free },
+    { OSSL_FUNC_KEYMGMT_GET_PARAMS, (void (*) (void))template_get_params },
+    { OSSL_FUNC_KEYMGMT_GETTABLE_PARAMS, (void (*) (void))template_gettable_params },
+    { OSSL_FUNC_KEYMGMT_SET_PARAMS, (void (*) (void))template_set_params },
+    { OSSL_FUNC_KEYMGMT_SETTABLE_PARAMS, (void (*) (void))template_settable_params },
+    { OSSL_FUNC_KEYMGMT_HAS, (void (*)(void))template_has },
+    { OSSL_FUNC_KEYMGMT_MATCH, (void (*)(void))template_match },
+    { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void))template_imexport_types },
+    { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (void (*)(void))template_imexport_types },
+    { OSSL_FUNC_KEYMGMT_IMPORT, (void (*)(void))template_export },
+    { OSSL_FUNC_KEYMGMT_EXPORT, (void (*)(void))template_import },
```

Comment:
Seems like these two entries should be the other way around

Reply:
embarassing. fixed. Thanks for the catch @mattcaswell !